File: lua/hold_state_machine.lua
local HoldStateMachine = {}
HoldStateMachine .__index = HoldStateMachine 

function HoldStateMachine.new(hold, blink)
  local result = {
    hold=hold or 0,
    blink=blink or 0,
    blink_start=nil,
    start=steady.clock(),
    blink_time = 0.0
  }
  setmetatable(result, HoldStateMachine)
  return result
end

function HoldStateMachine:lost()
  if self.blink_start == nil then
    self.blink_start = steady.clock()
  end
end

function HoldStateMachine:acquired(value)
  if value == nil then
    return self:acquired(true)
  end

  if not value then
    return self:lost()
  end

  if not self.blink_start then
    return
  end

  local now = steady.clock()
  self.blink_time = self.blink_time + now - self.blink_start
  self.blink_start = nil
end

function HoldStateMachine:update()
  local now = steady.clock()
  --print((now - self.start) .. (self.hold + self.blink_time))
  if self.blink_start ~= nil then
    local current_blink_time = now - self.blink_start
    --print(current_blink_time .. ' ' .. self.blink)
    if current_blink_time > self.blink then
      return "FAIL"
    end
  elseif now - self.start > self.hold + self.blink_time then
    return "SUCCESS"
  end
  return "PENDING"
end

return HoldStateMachine

File: src/hold_state_machine.cpp
#include <hold_state_machine.hpp>
#include <optional>

using namespace std::chrono_literals;

struct HoldStateMachine::Impl {
  std::chrono::milliseconds hold;
  std::chrono::milliseconds blink;
  std::optional<std::chrono::steady_clock::time_point> blink_start;
  std::chrono::steady_clock::time_point start;
  std::chrono::milliseconds blink_time = 0ms;

  Impl()
    : hold(0ms), blink(0ms), start(std::chrono::steady_clock::now()) {}
  
  Impl(std::chrono::milliseconds _hold, std::chrono::milliseconds _blink)
    : hold(_hold), blink(_blink), start(std::chrono::steady_clock::now()) {}
};

HoldStateMachine::HoldStateMachine(HoldStateMachine&& other) : impl(new Impl()) {
  std::swap(impl, other.impl);
}
HoldStateMachine::HoldStateMachine()
  : impl(new Impl()) {}

HoldStateMachine::HoldStateMachine(std::chrono::milliseconds hold, std::chrono::milliseconds blink)
  : impl(new Impl(hold, blink)) {}
HoldStateMachine::~HoldStateMachine() {}

HoldStateMachine& HoldStateMachine::operator=(HoldStateMachine&& other) {
  std::swap(impl, other.impl);
  return *this;
}

void HoldStateMachine::lost() {
  impl->blink_start = std::chrono::steady_clock::now();
}

void HoldStateMachine::acquired(bool value) {
  if(!value) {
    return lost();
  }
  if(!impl->blink_start.has_value()) {
    return;
  }
  auto now = std::chrono::steady_clock::now();
  impl->blink_time += std::chrono::duration_cast<std::chrono::milliseconds>(now - *impl->blink_start);
  impl->blink_start = std::nullopt;
}

HoldStateMachine::Status HoldStateMachine::update() {
  auto now = std::chrono::steady_clock::now();
  if(impl->blink_start.has_value()) {
    auto current_blink_time = now - *impl->blink_start;
    if(current_blink_time >= impl->blink) {
      return Status::FAIL;
    }
  } else if (now - impl->start > impl->hold + impl->blink_time) {
    return Status::SUCCESS;
  }
  return Status::PENDING;
}
